cpp-terminal/platforms/input.cpp

```diff
@@ -26,19 +45,167 @@ std::string to_utf8(LPCWCH utf16Str)
   WideCharToMultiByte(CP_UTF8, 0, utf16Str, 1, &ret[0], static_cast<int>(ret.capacity()), nullptr, nullptr);
   return ret.c_str();
 }
-#else
-namespace
+#endif
+
+namespace Private
 {
-volatile std::sig_atomic_t gSignalStatus;
-}
+template<typename T> class BlockingQueue
+{
+private:
+  std::mutex    mutex_;
+  std::queue<T> queue_;
+
+public:
+  T pop()
+  {
+    this->mutex_.lock();
+    T value = this->queue_.front();
+    this->queue_.pop();
+    this->mutex_.unlock();
+    return value;
+  }
+
+  void push(const T& value)
+  {
+    this->mutex_.lock();
+    this->queue_.push(value);
+    this->mutex_.unlock();
+  }
+
+  void push(T&& value)
+  {
+    this->mutex_.lock();
+    this->queue_.push(std::move(value));
+    this->mutex_.unlock();
+  }
+
+  bool empty()
+  {
+    this->mutex_.lock();
+    bool check = this->queue_.empty();
+    this->mutex_.unlock();
+    return check;
+  }
+  std::size_t size()
+  {
+    this->mutex_.lock();
+    std::size_t check = this->queue_.size();
+    this->mutex_.unlock();
+    return check;
+  }
+};
 
-static void sigwinchHandler(int sig)
+#if defined(__APPLE__) || defined(__wasm__) || defined(__wasm) || defined(__EMSCRIPTEN__)
+volatile std::sig_atomic_t m_signalStatus{0};
+static void                sigwinchHandler(int sig)
 {
-  if(sig == SIGWINCH) gSignalStatus = 1;
+  if(sig == SIGWINCH) m_signalStatus = 1;
 }
+#elif defined(_WIN32)
+#else
+class fd
+{
+public:
+  fd(const int& i) : m_fd(i) {}
+  ~fd() { ::close(m_fd); }
+  void set(const int& i) { m_fd = i; }
+  int  get() { return m_fd; }
+
+private:
+  int m_fd{-1};
+};
 #endif
 
-Term::Event Term::Platform::read_raw()
+}  // namespace Private
+}  // namespace Term
+
+std::thread Term::Private::Input::m_thread = std::thread(Term::Private::Input::read_event);
+
+Term::Private::BlockingQueue<Term::Event> Term::Private::Input::m_events;
+
+void Term::Private::Input::read_event()
+{
+  while(true)
+  {
+#if defined(_WIN32)
+    Term::Event ret;
+    WaitForSingleObject(Term::Private::in.handle(), INFINITE);
```

Comment:
Oh, nice so now we have an extra thread that actually brings the OS to wait for some input instead of polling every 10ns, right?

Reply:
Yes, and we ask the OS to hang until there is an event. I guess this would solve very fast typing who could be merged as one event.

Reply:
I have not yet used the c++ std threads a lot. When we pause one thread with a breaking system call, does only this thread pause or does the whole process pause? Do you have any information on that?

Reply:
I have changed the events.cpp in example. I add a thread printing some text to see if the thread is stopped. In linux and windows it seems to continue.